apps/api/models.py

```diff
@@ -102,24 +105,50 @@ class APIKey(models.Model):
     class Meta:
         db_table = 'api_key'
 
+    def __unicode__(self):
+        return (
+            u'<{cls} user={user}, type={type}, key={key} secret=...>'
+            .format(cls=self.__class__.__name__, key=self.key,
+                    type=self.type, user=self.user))
+
     @classmethod
     def get_jwt_key(cls, **query):
         """
-        Look up the config for a JWT key.
+        return a single APIKey instance for a JWT key matching the query.
         """
+        query.setdefault('is_active', True)
         return cls.objects.get(type=SYMMETRIC_JWT_TYPE, **query)
 
     @classmethod
     def new_jwt_credentials(cls, user, **attributes):
-        # This could get more fancy if we allow multiple keys per user.
-        key = 'user:{}'.format(user.pk)
+        """
+        Generates a new key/secret pair suitable for symmetric JWT signing.
+
+        This method must be run within a db transaction.
+        Returns an instance of APIKey.
+        """
+        key = cls.get_unique_key('user:{}:'.format(user.pk))
         return cls.objects.create(key=key, secret=cls.generate_secret(32),
                                   type=SYMMETRIC_JWT_TYPE, user=user,
-                                  **attributes)
+                                  is_active=True, **attributes)
+
+    @classmethod
+    def get_unique_key(cls, prefix, try_count=1, max_tries=1000):
+        if try_count >= max_tries:
+            raise RuntimeError(
+                'a unique API key could not be found after {} tries'
+                .format(max_tries))
+
+        key = '{}{}'.format(prefix, random.randint(0, 999))
+        if cls.objects.filter(key=key).exists():
+            return cls.get_unique_key(prefix, try_count=try_count + 1,
+                                      max_tries=max_tries)
```

Comment:
While I truly love some nice recursivity (even better when it's "tail recursive", with a [linear iterative process](https://mitpress.mit.edu/sicp/full-text/book/book-Z-H-11.html#%_sec_1.2.1)), I'm not sure I see the benefit over a simple for/else loop in this case.